waspc/data/Generator/templates/react-app/src/webSocket.ts

```diff
@@ -0,0 +1,32 @@
+import { useContext, useEffect } from 'react'
+import { WebSocketContext } from './webSocket/WebSocketProvider';
+
+import type {
+  ClientToServerEvents as _ClientToServerEvents,
+  ServerToClientEvents as _ServerToClientEvents
+} from '../../server/src/webSocket'
+
+export type ClientToServerEvents = _ClientToServerEvents
+export type ServerToClientEvents = _ServerToClientEvents
+
+export function useSocket() {
+  return useContext(WebSocketContext);
+}
+
+export function useSocketListener<Event extends keyof ServerToClientEvents>(
+  event: Event,
+  handler: ServerToClientEvents[Event],
+) {
+  const { socket } = useContext(WebSocketContext);
+  useEffect(() => {
+    // TODO(miho): This is a hack for type errors we are getting
+    // from socket.on when passing in "handler" directly
+    const handlerInstance: any = (event: any) => {
+      handler(event);
+    };
```

Comment:
Can we elaborate on the error and list a proper fix (if we know it)?

Reply:
I get the following error:
```
Argument of type 'ServerToClientEvents[Event]' is not assignable to parameter of type 'FallbackToUntypedListener<Event extends "disconnect" | "connect" | "connect_error" ? SocketReservedEvents[Event] : Event extends "chatMessage" ? ServerToClientEvents[Event] : never>'.
  Type '(msg: { id: string; username: string; text: string; }) => void' is not assignable to type 'FallbackToUntypedListener<Event extends "disconnect" | "connect" | "connect_error" ? SocketReservedEvents[Event] : Event extends "chatMessage" ? ServerToClientEvents[Event] : never>'.
```
I haven't been able to figure out why that happens after digging through socket-io's source code.

And I haven't been able to find a fully working Typescript socket.io example, but I did find this workaround.

Reply:
![Screenshot 2023-06-16 at 12 22 53](https://github.com/wasp-lang/wasp/assets/2223680/a0fd1fcc-8aa9-4612-b1a8-05013b8cae76)

Reply:
I've added a more detailed comment 👍

Reply:
How come you didn't use `as keyof ServerToClientEvents` in the end?

Reply:
I've forgot about that change 🤦‍♂️ I did it in the generated code and not in the template.

Reply:
Updated the code to use the cast 👍